### PR TITLE
Feature/dynamic select parameter component for report builder

### PR DESCRIPTION
**Pull request by DXSaurabh**
**_NOTE: @davidjgonzalez made some suggestions/changes via #1258 but cannot PR it to DXSaurabh so i just PR'd it to master. This and #1258 are ~ the same, and eventually one will need to supersede the other._**

Hi @justinedelson/ @dchang ,

I created the one new component select parameter for report builder component. This component is having the feature for populating the select parameter drop-down by JCR Query instead of static options.

dynamic-dropdown

Thanks

### CODE REVIEW

File: content/src/main/content/jcr_root/apps/acs-commons/components/utilities/report-builder/parameters/dynamic-select/datasource/datasource.jsp [hunk 1]
```diff
@@ -0,0 +1,86 @@
+<%@page session="false" import="
```
Comment (davidjgonzalez) [inline content/src/main/content/jcr_root/apps/acs-commons/components/utilities/report-builder/parameters/dynamic-select/datasource/datasource.jsp]: hi @DXSaurabh , cany chance you could port this over to a SlingSafeMethodsServlet?

We have a handy DataSourceBuilder API [1] in ACS Commons that can help w some of this boilerplate code. 

Recognizing the irony that im using a JSP as a reference for how to use this API :) ..you can check out [2]

It almost might be cool to add another field that lets a user specify a Query languages and you use ACS's QueryHelper [3] to construct and execute the query in any of the supported languages -- though it looks like youre parsing queries by hand? ..  



[1] https://github.com/Adobe-Consulting-Services/acs-aem-commons/blob/master/bundle/src/main/java/com/adobe/acs/commons/wcm/datasources/DataSourceBuilder.java
[2] https://github.com/Adobe-Consulting-Services/acs-aem-commons/blob/master/content/src/main/content/jcr_root/apps/acs-commons/datasources/named-image-transforms/named-image-transforms.jsp#L27
[3] https://github.com/Adobe-Consulting-Services/acs-aem-commons/blob/master/bundle/src/main/java/com/adobe/acs/commons/util/QueryHelper.java

File: content/src/main/content/jcr_root/apps/acs-commons/components/utilities/report-builder/parameters/dynamic-select/datasource/datasource.jsp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                        } 
+                    }//End of while parsing the property           
+            } 
+    //Create a DataSource that is used to populate the drop-down control
+    DataSource ds = new SimpleDataSource(fakeResourceList.iterator());
+    request.setAttribute(DataSource.class.getName(), ds);
+%>
+<%! 
+    //Mehtod for selecting the column from query
+   public List<String> parseQuery(String jcrQueryString) {
```
Comment (davidjgonzalez) [inline content/src/main/content/jcr_root/apps/acs-commons/components/utilities/report-builder/parameters/dynamic-select/datasource/datasource.jsp]: I dont quite understand the purpose of this..  what do the query columns / checks do here?

File: content/src/main/content/jcr_root/apps/acs-commons/components/utilities/report-builder/parameters/dynamic-select/_cq_dialog/.content.xml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                    <datasource
+                        jcr:primaryType="nt:unstructured"
+                        sling:resourceType="granite/ui/components/coral/foundation/form/textfield"
+                        fieldLabel="Datasource"
+                        name="./datasource/sling:resourceType"/>
+                    <dropdownquery
+                        jcr:primaryType="nt:unstructured"
+                        sling:resourceType="granite/ui/components/coral/foundation/form/textarea"
+                        fieldLabel="Drop-down Query"
+                        name="./dropdownquery"/>
```
Comment (davidjgonzalez) [inline content/src/main/content/jcr_root/apps/acs-commons/components/utilities/report-builder/parameters/dynamic-select/_cq_dialog/.content.xml]: jcr props should be camelCase

File: content/src/main/content/jcr_root/apps/acs-commons/components/utilities/report-builder/parameters/dynamic-select/_cq_dialog/.content.xml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+                        value="granite/ui/components/coral/foundation/form/select"/>
+                    <datasource
+                        jcr:primaryType="nt:unstructured"
+                        sling:resourceType="granite/ui/components/coral/foundation/form/textfield"
+                        fieldLabel="Datasource"
+                        name="./datasource/sling:resourceType"/>
+                    <dropdownquery
+                        jcr:primaryType="nt:unstructured"
+                        sling:resourceType="granite/ui/components/coral/foundation/form/textarea"
+                        fieldLabel="Drop-down Query"
```
Comment (davidjgonzalez) [inline content/src/main/content/jcr_root/apps/acs-commons/components/utilities/report-builder/parameters/dynamic-select/_cq_dialog/.content.xml]: Not sure that Drop-down is hypenated? I think its one word?

File: content/src/main/content/jcr_root/apps/acs-commons/components/utilities/report-builder/parameters/dynamic-select/datasource/datasource.jsp (hunk 1 above)
Comment (DXSaurabh) [reply content/src/main/content/jcr_root/apps/acs-commons/components/utilities/report-builder/parameters/dynamic-select/datasource/datasource.jsp]: Hi @davidjgonzalez , 

Thanks for reviewing the code.
I have ported the dataSource.jsp over to a SlingSafeMethodsServlet. Adding the one more field for query type is really a good idea but i can add this thing in commit of the new feature.

Thanks

File: content/src/main/content/jcr_root/apps/acs-commons/components/utilities/report-builder/parameters/dynamic-select/_cq_dialog/.content.xml (hunk 3 above)
Comment (DXSaurabh) [reply content/src/main/content/jcr_root/apps/acs-commons/components/utilities/report-builder/parameters/dynamic-select/_cq_dialog/.content.xml]: Done.

File: content/src/main/content/jcr_root/apps/acs-commons/components/utilities/report-builder/parameters/dynamic-select/_cq_dialog/.content.xml (hunk 4 above)
Comment (DXSaurabh) [reply content/src/main/content/jcr_root/apps/acs-commons/components/utilities/report-builder/parameters/dynamic-select/_cq_dialog/.content.xml]: Done.

**Comment by DXSaurabh** [comment]
Hi @davidjgonzalez 

I have added the JUnit class and tested the functionality with all query type options.

Can you please merge this feature now?

Kindly let me know if there is any suggestion!

Thanks
Saurabh

**Pull request merged**